Tuple Slice with both negative and positive stepped size

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#48660 Tuple Slice with both negative and positive stepped size**

Summary: We used to support tuple slicing without any step size before, but this PR extends this feature to support arbitrary step size. We do this by manually reconstructing a new tuple in the IR instead of relying on TupleSlice prim.

Test Plan: python tests

Reviewers: @gmagogsfm

Subscribers:

Tasks: https://github.com/pytorch/pytorch/issues/47334

Tags:

Differential Revision: [D25359336](https://our.internmc.facebook.com/intern/diff/D25359336)

tugsbayasgalan committed:
Tuple Slice with both negative and positive stepped size

Summary: We used to support tuple slicing without any step size before, but this PR extends this feature to support arbitrary step size. We do this by manually reconstructing a new tuple in the IR instead of relying on TupleSlice prim.

Test Plan: python tests

Reviewers: @gmagogsfm

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Tuple Slice with both negative and positive stepped size"


Summary: We used to support tuple slicing without any step size before, but this PR extends this feature to support arbitrary step size. We do this by manually reconstructing a new tuple in the IR instead of relying on TupleSlice prim.

Test Plan: python tests

Reviewers: @gmagogsfm

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

tugsbayasgalan commented:
commented out the TUPLE_SLICE previous logic for the visibility. If approved, I will go ahead and remove them

tugsbayasgalan committed:
Update on "Tuple Slice with both negative and positive stepped size"


Summary: We used to support tuple slicing without any step size before, but this PR extends this feature to support arbitrary step size. We do this by manually reconstructing a new tuple in the IR instead of relying on TupleSlice prim.

Test Plan: python tests

Reviewers: @gmagogsfm

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Tuple Slice with both negative and positive stepped size"


Summary: We used to support tuple slicing without any step size before, but this PR extends this feature to support arbitrary step size. We do this by manually reconstructing a new tuple in the IR instead of relying on TupleSlice prim.

Test Plan: python tests

Reviewers: @gmagogsfm

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3453,17 +3453,36 @@ struct to_ir {
-        // TODO: add support for slicing tuples with a step
-        throw ErrorReport(loc)
-            << "Unsupported operation: slicing tuples with a step isn't supported";
+      if (start && end && step) {
```
I think logic in this method can be simplified.

You can change type of `args` to `std::vector<at::optional<NamedValue>>`, then construct its 4 values according to nullness of start/end/step. This way you only need to make one call to `emitTupleSlice`.

tugsbayasgalan replied:
Good idea! will do

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3827,28 +3846,47 @@ struct to_ir {
+    int64_t step_size = 1;
+    if (step) {
+      auto val = toIValue(step->value(*graph));
+      if (val->isInt()) {
```
I think it is safe to `TORCH_CHECK(val->isInt())` here. Otherwise, step is silently ignored if a non-int value is provided.

tugsbayasgalan replied:
sounds good!

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3827,28 +3846,47 @@ struct to_ir {
+          loc, tuple_type, getSliceInd(beg_val->value(*graph), loc), true);
+    } else {
+      // we want to flip the beg if we are slicing in the opposite direction
+      beg = step_size > 0 ? (int64_t)0 : tuple_len;
```
Nit:
 
You probably don't need to cast 0 to int64_t. It should happen automatically.

gmagogsfm commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -870,12 +870,12 @@ struct CodeImpl {
-    int64_t end_ind = node->i(attr::end);
-    insertInstruction(TUPLE_SLICE, beg_ind, end_ind - beg_ind);
-  }
+  // void emitTupleSlice(Node* node) {
```
Remove?

gmagogsfm commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -973,9 +973,9 @@ struct CodeImpl {
-      case prim::TupleSlice:
-        emitTupleSlice(node);
-        break;
+      // case prim::TupleSlice:
```
Also remove

gmagogsfm commented on torch/csrc/jit/ir/ir.cpp:
```diff
@@ -1608,17 +1608,25 @@ Node* Graph::createTupleIndex(
+    int64_t beg,
+    int64_t end,
+    int64_t step_size) {
+  auto unpacked_tuple = insertNode(createTupleUnpack(tup));
```
I don't think you need to use `TupleUnpack` here, which reads out all values from tuple. Instead, I think `prim::TupleIndex` is better.

gmagogsfm commented on torch/csrc/jit/runtime/instruction.h:
```diff
@@ -50,7 +50,6 @@ namespace jit {
   _(DICT_CONSTRUCT, "TI") /* construct a dict of type X, using N inputs */     \
   _(CREATE_OBJECT, "T") /* create an object of type X */                       \
   _(ISINSTANCE, "TI") /* check object is one of  types[X:X+N]  */              \
-  _(TUPLE_SLICE, "II") /* slice tup[X:(X+N)] */                                \
```
We should probably talk with Martin Yuan on mobile team to see if he needs to mimic this change.

tugsbayasgalan replied:
@iseeyuan could you take a look?

iseeyuan replied:
It might affect the backward compatibility of mobile. If an old mobile model with TUPLE_SLICE is loaded by this code, it will fail. If the model is already deployed by production, there would be a regress. Is it possible to keep TUPLE_SLICE to make it backward compatible?

tugsbayasgalan committed:
Update on "Tuple Slice with both negative and positive stepped size"


Summary: We used to support tuple slicing without any step size before, but this PR extends this feature to support arbitrary step size. We do this by manually reconstructing a new tuple in the IR instead of relying on TupleSlice prim.

Test Plan: python tests

Reviewers: @gmagogsfm

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Tuple Slice with both negative and positive stepped size"


Summary: We used to support tuple slicing without any step size before, but this PR extends this feature to support arbitrary step size. We do this by manually reconstructing a new tuple in the IR instead of relying on TupleSlice prim.

Test Plan: python tests

Reviewers: @gmagogsfm

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3827,28 +3827,47 @@ struct to_ir {
+    auto beg_val = tuple_args[0];
+    auto end_val = tuple_args[1];
+    auto step = tuple_args[2];
+    // first figure out the step sign to know if we want to flip the start or
```
Nit:
`start or end` -> `start and end`

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3827,28 +3827,47 @@ struct to_ir {
+          loc, tuple_type, getSliceInd(beg_val->value(*graph), loc), true);
+    } else {
+      // we want to flip the beg if we are slicing in the opposite direction
+      beg = step_size > 0 ? 0 : tuple_len;
```
Should it be `0` or `tuple_len` - 1? IIRC, `begin` is inclusive so it must be a valid index.

Also none of your test cases cover this case (negative step, begin not specified)

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3827,28 +3827,47 @@ struct to_ir {
     } else {
-      end = tuple_len;
+      // we want to flip the end if we are slicing in the opposite direction
+      end = step_size > 0 ? tuple_len : 0;
```
Similar situation here. `end` is not-inclusive, so it should be tuple_len or -1 I think.

gmagogsfm commented on torch/csrc/jit/ir/ir.cpp:
```diff
@@ -1608,17 +1608,30 @@ Node* Graph::createTupleIndex(
+  std::vector<Value*> newValues;
+  TupleTypePtr tt = tup->type()->expect<TupleType>();
+
+  if (step_size > 0) {
```
Now that I think about it. It is a bit confusing if we flip `begin` and `end` earlier but now always iterate from begin and end, but with a different loop condition `i > end` vs `i < end` here. IMHO, this logic would be more readable if we put all logic about checking step-size being positive or negative then perform different actions here. Let me know what you think.

tugsbayasgalan replied:
yeah that is a good idea. I will try to move the logic into one step

tugsbayasgalan commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3453,17 +3453,17 @@ struct to_ir {
-      } else {
-        return emitTupleSlice(loc, args[0], args[1], c10::nullopt);
-      }
+      start ? tuple_args.emplace_back(start)
```
Seems like both options need to be same type for this to work. I tried this before and was getting this "error: operands to ?: have different types ‘torch::jit::Value*’ and ‘const c10::nullopt_t’". I feel like fixing this would make the logic little bit more complicated than necessary.

tugsbayasgalan committed:
Update on "Tuple Slice with both negative and positive stepped size"


Summary: We used to support tuple slicing without any step size before, but this PR extends this feature to support arbitrary step size. We do this by manually reconstructing a new tuple in the IR instead of relying on TupleSlice prim.

Test Plan: python tests

Reviewers: @gmagogsfm

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Tuple Slice with both negative and positive stepped size"


Summary: We used to support tuple slicing without any step size before, but this PR extends this feature to support arbitrary step size. We do this by manually reconstructing a new tuple in the IR instead of relying on TupleSlice prim.

Test Plan: python tests

Reviewers: @gmagogsfm

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Tuple Slice with both negative and positive stepped size"


Summary: We used to support tuple slicing without any step size before, but this PR extends this feature to support arbitrary step size. We do this by manually reconstructing a new tuple in the IR instead of relying on TupleSlice prim.

Test Plan: python tests

Reviewers: @gmagogsfm

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/runtime/slice_indices_adjust.cpp:
```diff
@@ -0,0 +1,56 @@
+namespace torch {
+namespace jit {
+
+int64_t PySlice_AdjustIndices(
```
Though this is copied from Python, we can change its name to something more appropriate for our use case right?

gmagogsfm commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -3826,28 +3832,38 @@ struct to_ir {
+          loc, tuple_type, getSliceInd(beg_val->value(*graph), loc), true);
+    }
+
+    int64_t end = INT64_MAX;
```
same here

tugsbayasgalan committed:
Update on "Tuple Slice with both negative and positive stepped size"


Summary: We used to support tuple slicing without any step size before, but this PR extends this feature to support arbitrary step size. We do this by manually reconstructing a new tuple in the IR instead of relying on TupleSlice prim.

Test Plan: python tests

Reviewers: @gmagogsfm

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

gmagogsfm reviewed (approved):
Looks good! Thanks for the fast iterations. There are still a few broken tests according to CI, please make sure they are fixed before you land.

tugsbayasgalan committed:
Update on "Tuple Slice with both negative and positive stepped size"


Summary: We used to support tuple slicing without any step size before, but this PR extends this feature to support arbitrary step size. We do this by manually reconstructing a new tuple in the IR instead of relying on TupleSlice prim.

Test Plan: python tests

Reviewers: @gmagogsfm

Subscribers:

Tasks:

Tags:

Differential Revision: [D25359336](https://our.internmc.facebook.com/intern/diff/D25359336)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Tuple Slice with both negative and positive stepped size"


Summary: We used to support tuple slicing without any step size before, but this PR extends this feature to support arbitrary step size. We do this by manually reconstructing a new tuple in the IR instead of relying on TupleSlice prim.

Test Plan: python tests

Reviewers: @gmagogsfm

Subscribers:

Tasks:

Tags:

Differential Revision: [D25359336](https://our.internmc.facebook.com/intern/diff/D25359336)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Tuple Slice with both negative and positive stepped size"


Summary: We used to support tuple slicing without any step size before, but this PR extends this feature to support arbitrary step size. We do this by manually reconstructing a new tuple in the IR instead of relying on TupleSlice prim.

Test Plan: python tests

Reviewers: @gmagogsfm

Subscribers:

Tasks:

Tags:

Differential Revision: [D25359336](https://our.internmc.facebook.com/intern/diff/D25359336)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Tuple Slice with both negative and positive stepped size"


Summary: We used to support tuple slicing without any step size before, but this PR extends this feature to support arbitrary step size. We do this by manually reconstructing a new tuple in the IR instead of relying on TupleSlice prim.

Test Plan: python tests

Reviewers: @gmagogsfm

Subscribers:

Tasks:

Tags:

Differential Revision: [D25359336](https://our.internmc.facebook.com/intern/diff/D25359336)

[ghstack-poisoned]
